### PR TITLE
feature: BOTI evaluation implementation

### PR DESCRIPTION
**Pull request by IskandarGhobril**
This PR introduces the evaluation logic, UI for the **Battle of the Interviewers (BOTI)** feature, and the corresponding tests.

#### Key Features:
- **Evaluation Process:**  
   The evaluation begins as soon as both users finish their interaction and access the evaluation screen. The challenger's side initiates the evaluation, and the results are uploaded to Firebase to be shared with the opponent. ChatGPT analyzes the full message history of both users, simulating a strict recruiter to:
   - Decide the winner.
   - Provide constructive feedback tailored for both the winner and the loser.

- **Evaluation Logic Implementation:**  
   - **`evaluateBattleCandidates`:**  
     - Formats chat transcripts for both users and queries ChatGPT to determine the winner.
     - Extracts the winner's UID and feedback for both participants using regex from ChatGPT's response.  
   - **`completeBattle`:** Updates the battle's status in Firestore to `COMPLETED` and stores the evaluation results, making them accessible to both users.

- **Evaluation Screen Implementation:**  
   The evaluation screen transitions through multiple states dynamically:
   - **Loading Data:** While fetching battle data from Firestore.
   - **Evaluating Performance:** Displays a progress indicator during the evaluation process.  
     eval performance
   - **Displaying Results:** Shows the outcome and detailed feedback for each user.
   - **Action Buttons:** Once the result is displayed, the screen provides buttons to:
     - Retry the battle.
     - Navigate to the Home screen.
     - Start an Interview Practice session.  
     eval done

#### Next Steps for BOTI:
- Adding a **cancellation mechanism** to handle cases where users want to exit the battle.
- Implementing a **retry mechanism** to allow users to replay battles with the same opponent.
- Moving the battle initiation flow to a new dedicated screen, streamlining the user experience.

### CODE REVIEW

File: app/src/main/java/com/github/se/orator/model/chatGPT/ChatViewModel.kt [hunk 1]
```diff
@@ -516,7 +517,117 @@ The session is now over. According to the initial instructions, you can now brea
...
+      user2Uid: String,
+      user1messages: List<Message>,
+      user2messages: List<Message>,
+      user1name: String,
+      user2name: String
+  ): EvaluationResult {
+    val user1Transcript = messagesToTranscript(user1Uid, user1messages)
+    val user2Transcript = messagesToTranscript(user2Uid, user2messages)
+
+    val prompt =
```
Comment (baptisteghostine) [inline app/src/main/java/com/github/se/orator/model/chatGPT/ChatViewModel.kt:552]: #### **Enhance Prompt Maintainability in ChatViewModel**

Great work on integrating the evaluation logic! To make the ChatGPT prompts in `ChatViewModel` more maintainable and easier to update, consider extracting them into a separate `PromptTemplates` object or a dedicated constants file. This will help keep your `ChatViewModel` clean and make it simpler to manage and modify prompts in the future.

---

#### **Example Implementation**

Create a new Kotlin object to house all your prompt templates:

```kotlin
// PromptTemplates.kt
object PromptTemplates {
    fun evaluationPrompt(
        user1Uid: String,
        user2Uid: String,
        user1Transcript: String,
        user2Transcript: String
    ): String {
        return """
            You are a strict recruiter deciding between $user1Uid and $user2Uid.

            Q&A transcripts:

            $user1Uid's transcript:
            $user1Transcript

            $user2Uid's transcript:
            $user2Transcript

            Determine the winner and explain briefly why. Also, give concise feedback for the loser.
            Format:
            The winner is: <winnerUid>
            Winner message: <short message>
            Loser message: <short message>
        """.trimIndent()
    }
}
```

Update the evaluateBattleCandidates function in ChatViewModel to utilize this template:

```kotlin
val prompt = PromptTemplates.evaluationPrompt(
    user1Uid = battle.challenger,
    user2Uid = battle.opponent,
    user1Transcript = user1Transcript,
    user2Transcript = user2Transcript
)
```

File: app/src/main/java/com/github/se/orator/model/chatGPT/ChatViewModel.kt [hunk 2]
```diff
@@ -516,7 +517,117 @@ The session is now over. According to the initial instructions, you can now brea
...
+        
+        $user1Uid's transcript:
+        $user1Transcript
+
+        $user2Uid's transcript:
+        $user2Transcript
+
+        Please determine who performed better and explicitly state: "The winner is $user1Uid" or "The winner is $user2Uid".
+        Then, provide a message addressed to the winner, highlighting why they won (max one paragraph). 
+        Also, provide a message addressed to the loser, highlighting what they did wrong and where they can improve (max one paragraph).
```
Comment (baptisteghostine) [inline app/src/main/java/com/github/se/orator/model/chatGPT/ChatViewModel.kt:566]: Maybe a better prompt would be : 

```plaintext
You are an impartial and strict recruiter tasked with evaluating two AI interview candidates, $user1Uid and $user2Uid, based on their interview transcripts. Your goal is to determine which AI performed better and provide constructive feedback for both.

            **Guidelines:**

            1. **Evaluation Criteria:**
               - **Clarity and Coherence:** Assess how clearly and logically each AI communicated their answers.
               - **Depth of Responses:** Evaluate the thoroughness and depth of each AI's answers.
               - **Relevance to Questions:** Determine how well each AI addressed the interview questions.
               - **Problem-Solving Ability:** Analyze the effectiveness of each AI's problem-solving approach.
               - **Professionalism:** Consider the professionalism displayed in each AI's responses.

            2. **Determining the Winner:**
               - Based on the criteria above, decide which AI performed better overall.
               - Explicitly state the winner in the format: "The winner is: <winnerUid>."

            3. **Feedback for Both AIs:**
               - **Winner's Feedback:** Provide specific strengths that led to their victory.
               - **Loser's Feedback:** Offer constructive criticism highlighting areas for improvement.

            **Format your response as follows:**

            ```
            The winner is: <winnerUid>
            Winner's Feedback:
            - <Strength 1>
            - <Strength 2>
            - <Strength 3>

            Loser's Feedback:
            - <Area for Improvement 1>
            - <Area for Improvement 2>
            - <Area for Improvement 3>
            ```
            
            **Example:**

            ```
            The winner is: user1Uid
            Winner's Feedback:
            - Demonstrated exceptional problem-solving skills with clear, logical explanations.
            - Provided comprehensive and relevant answers to all questions.
            - Maintained a high level of professionalism throughout the interview.

            Loser's Feedback:
            - Responses lacked depth and were sometimes vague.
            - Failed to fully address some of the interview questions.
            - Could improve clarity and coherence in communication.
            ```

            **Transcripts:**

            **$user1Uid's Transcript:**
            $user1Transcript

            **$user2Uid's Transcript:**
            $user2Transcript
        """.trimIndent()
    }
}
```

File: app/src/main/java/com/github/se/orator/model/speechBattle/BattleViewModel.kt [hunk 3]
```diff
@@ -277,4 +279,45 @@ class BattleViewModel(
       }
     }
   }
+
+  fun evaluateBattle(battleId: String, onFailure: (Exception) -> Unit) {
```
Comment (baptisteghostine) [inline app/src/main/java/com/github/se/orator/model/speechBattle/BattleViewModel.kt:283]: I really like how you've implemented the evaluation logic! To make debugging even easier, let's enhance the error handling in the `evaluateBattle` function by adding more context to the log messages. Including identifiers like `battleId`, `user1Uid`, and `user2Uid` will make tracing issues much simpler.

In your `BattleViewModel.kt`, update the `evaluateBattle` function's error handling as follows:

```kotlin

suspend fun evaluateBattle(battleId: String, onFailure: (Exception) -> Unit) {
    val currentUserUid = userProfileViewModel.userProfile.value?.uid ?: return

    getBattleById(battleId) { battle ->
        if (battle == null) {
            onFailure(IllegalStateException("Battle not found for battleId=$battleId"))
            return@getBattleById
        }

        // Only challenger triggers the evaluation
        if (battle.challenger != currentUserUid) {
            // This user is not the challenger, do nothing special here.
            return@getBattleById
        }

        // Evaluate the battle candidates
        viewModelScope.launch {
            try {
                val evaluationResult = chatViewModel.evaluateBattleCandidates(
                    user1Uid = battle.challenger,
                    user2Uid = battle.opponent,
                    user1messages = battle.challengerData,
                    user2messages = battle.opponentData,
                    user1name = userProfileViewModel.getName(battle.challenger),
                    user2name = userProfileViewModel.getName(battle.opponent)
                )

                battleRepository.completeBattle(battleId, evaluationResult) { success ->
                    if (!success) {

**Review by baptisteghostine — CHANGES_REQUESTED** [review]
#### **PR Review: Evaluation Logic for Battle of the Interviewers (BOTI)**

#### **Summary**
This PR introduces evaluation logic for the **Battle of the Interviewers (BOTI)** feature, with enhancements to the `ChatViewModel`, `BattleViewModel`, and the `EvaluationScreen`. Key updates include candidate evaluation methods, Firestore integration for syncing battle results, and dynamic UI states for displaying outcomes. These changes significantly improve the user experience and set the groundwork for future BOTI iterations.

A little note before starting: Your PR is quite heavy in terms of lines. Maybe split it in 2 for next time so it is easier to do a PR Review. 

#### **Highlights of the PR**

1. **Evaluation Flow Integration**
   - Introduced the `evaluateBattleCandidates` function to determine the winner and generate contextual feedback using ChatGPT.
   - Seamlessly integrates the evaluation flow into the battle lifecycle.

2. **Firestore Updates**
   - Added methods like `completeBattle` and `evaluateBattleCandidates` to update battle status and store evaluation results, ensuring both participants remain synchronized.

3. **Dynamic UI States**
   - The `EvaluationScreen` now handles `loading`, `evaluating`, and `completed` states, providing a smoother and more informative user experience.

#### **Suggestions and Requested Changes**

##### **1. Prompt Quality and Maintainability**
- **Comment**: The ChatGPT evaluation prompts are great but it should be refined to appear less hardcoded and be more modular.
- **Suggestion**:
  - Extract prompts into reusable, well-structured constants or functions.
  - Use placeholders for roles, performance criteria, and other parameters to simplify future updates.
- **Example**:
  - Move prompts to a separate file, such as `EvaluationPrompts.kt`, and structure them like:
    ```kotlin
    const val EVALUATION_PROMPT = """
    Evaluate the candidates based on the following criteria:
    - Role: {role}
    - Performance Criteria: {criteria}
    Provide detailed feedback and determine a winner.
    """
    ```
  - This ensures flexibility and consistency in how prompts are applied across the app.

##### **2. Error Handling and Logging**
- **Comment**: Exceptions in `ChatViewModel` and `BattleViewModel` lack descriptive error messages.
- **Suggestion**:
  - Log exceptions with more context to simplify debugging. Include details like battle ID, candidate IDs, and the prompt used in ChatGPT requests.
  - Example:
    ```kotlin
    Log.e("BattleViewModel", "Failed to evaluate candidates for battleId: $battleId. Error: ${e.message}")
    ```

##### **3. Code Organization and Reusability**
- **Comment**: Some logic, such as assembling transcripts or parsing user feedback, is repetitive.
- **Suggestion**:
  - Extract repetitive logic into utility functions or helper classes to improve maintainability.
  - Separate ChatGPT request logic and regex parsing into reusable components for easier testing and debugging.

##### **4. Testing and Validation**
- **Comment**: Testing focuses primarily on success scenarios, with limited coverage for edge cases.
- **Suggestions**:
  - Add tests for failure scenarios, such as:
    - ChatGPT returning incomplete or invalid data.
    - Regex failing to match expected patterns.
  - Test all UI state transitions in `EvaluationScreen`, ensuring smooth progression from `loading` to `evaluation` and `completed` states.

#### **Final Feedback**
This PR represents a significant enhancement for the BOTI feature, delivering meaningful evaluation logic and a polished user flow. By refining prompt structure, improving error handling, and increasing test coverage, this PR can establish a solid foundation for future iterations of BOTI. Excellent work—keep it up!

File: app/src/main/java/com/github/se/orator/model/chatGPT/ChatViewModel.kt (hunk 2 above)
Comment (IskandarGhobril) [reply app/src/main/java/com/github/se/orator/model/chatGPT/ChatViewModel.kt:566]: Thank you for the suggestion! It's a great improvement compared to my version

File: app/src/main/java/com/github/se/orator/model/chatGPT/ChatViewModel.kt (hunk 1 above)
Comment (IskandarGhobril) [reply app/src/main/java/com/github/se/orator/model/chatGPT/ChatViewModel.kt:552]: This is a great suggestion. However, it would be better to do this for all the prompts in a future PR, as it’s currently not implemented for the other prompts in the chat view model either.

File: app/src/main/java/com/github/se/orator/model/speechBattle/BattleViewModel.kt (hunk 3 above)
Comment (IskandarGhobril) [reply app/src/main/java/com/github/se/orator/model/speechBattle/BattleViewModel.kt:283]: Thanks for the suggestion!

**Comment by baptisteghostine** [comment]
Thank you for addressing my request carefully, the prompts are really great now, and this PR adds good value to the app. Proud of your work! I will now proceed to accept the PR! Thank you and have a great evening.

**Review by baptisteghostine — APPROVED** [review]
